sdk/pinocchio/src/instruction.rs

```diff
@@ -78,6 +81,26 @@ const fn offset<T, U>(ptr: *const T, offset: usize) -> *const U {
     unsafe { (ptr as *const u8).add(offset) as *const U }
 }
 
+impl Account<'_> {
+    /// Return true if the account borrow state is set to the given state.
+    ///
+    /// This will test both data and lamports borrow state, similarly to
+    /// [`AccountInfo::is_borrowed`].
+    #[inline(always)]
+    pub fn is_borrowed(&self, state: BorrowState) -> bool {
+        // SAFETY: The `borrow_state` of an account is stored in the first byte of the
+        // memory location of the account on the input buffer, which starts at 8 bytes
+        // before the account `key`.
+        unsafe { *(self.key as *const u8).sub(8) & (state as u8) != 0 }
```

Comment:
This is extremely unsafe, and only works in a VM context, where `Account` was created from an `AccountInfo` passed by the runtime. If we can avoid having this function at all and instead perform the checks on the `AccountInfo`s, it would definitely be safer.

We might just want to have the `invoke_signed` for those who want some validation checks, and `invoke_instruction_signed_unchecked` who want full control. Anything in between might not be necessary and more confusing than helpful

Reply:
I have done a bit more testing with the `invoke_signed` and it seems that there is almost no difference compared to the proposed `invoke_instruction_signed`. I think it is ok to remove the new ones and keep the existing API.